NIFI-8294 Add Query Processor for Azure Data Explorer

tanmaya-panda1 opened the pull request:
<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

[NIFI-8294](https://issues.apache.org/jira/browse/NIFI-8294)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [ ] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [ ] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ ] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [ ] Pull Request based on current revision of the `main` branch
- [ ] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [ ] JDK 11
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-api/pom.xml:
```diff
@@ -0,0 +1,89 @@
+    </dependencies>
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.jacoco</groupId>
+                <artifactId>jacoco-maven-plugin</artifactId>
+                <version>0.8.6</version>
+                <executions>
+                    <execution>
+                        <id>prepare-agent</id>
+                        <goals>
+                            <goal>prepare-agent</goal>
+                        </goals>
+                    </execution>
+                    <execution>
+                        <id>report</id>
+                        <phase>test</phase>
+                        <goals>
+                            <goal>report</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
```
The code coverage plugin configuration should be removed. This is something that can be considered at a project level, but it should not be configured for specific modules.

exceptionfactory replied:
```suggestion
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-api/pom.xml:
```diff
@@ -0,0 +1,89 @@
+                    </excludes>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <configuration>
+                    <systemPropertyVariables>
+                        <java.io.tmpdir>${project.build.directory}</java.io.tmpdir>
+                    </systemPropertyVariables>
+                </configuration>
+            </plugin>
```
This configuration should be removed. The standard temporary directory should be used without changing it to the project build directory.

exceptionfactory replied:
```suggestion
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-nar/pom.xml:
```diff
@@ -0,0 +1,63 @@
+                    </excludes>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <configuration>
+                    <systemPropertyVariables>
+                        <java.io.tmpdir>${project.build.directory}</java.io.tmpdir>
+                    </systemPropertyVariables>
+                </configuration>
+            </plugin>
+        </plugins>
```
```suggestion
```

exceptionfactory replied:
Is there a specific reason for changing the temporary directory? This seems unnecessary.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-nar/pom.xml:
```diff
@@ -0,0 +1,63 @@
+
+    <artifactId>nifi-adx-nar</artifactId>
+    <packaging>nar</packaging>
+    <properties>
+        <maven.javadoc.skip>true</maven.javadoc.skip>
+        <source.skip>true</source.skip>
+    </properties>
```
This options can be removed.
```suggestion
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/pom.xml:
```diff
@@ -0,0 +1,126 @@
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-adx</artifactId>
+            <version>${nifi-release-version}</version>
```
Using this `nifi-release-version` is not a standard across the repository. It should be replaced with the explicit version `2.0.0-SNAPSHOT`.
```suggestion
            <version>2.0.0-SNAPSHOT</version>
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/pom.xml:
```diff
@@ -0,0 +1,126 @@
+            <artifactId>jcl-over-slf4j</artifactId>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>log4j-over-slf4j</artifactId>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>jul-to-slf4j</artifactId>
+            <scope>compile</scope>
+        </dependency>
```
These dependencies are applied at the project level and do not need to be declared for individual modules.
```suggestion
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/pom.xml:
```diff
@@ -0,0 +1,126 @@
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <version>5.9.0</version>
```
The version for JUnit is set at the project level, so this explicit version should be removed.
```suggestion
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/pom.xml:
```diff
@@ -0,0 +1,126 @@
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-params</artifactId>
+            <version>5.9.0</version>
```
```suggestion
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/pom.xml:
```diff
@@ -0,0 +1,126 @@
+    </dependencies>
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.jacoco</groupId>
+                <artifactId>jacoco-maven-plugin</artifactId>
+                <version>0.8.6</version>
+                <executions>
+                    <execution>
+                        <id>prepare-agent</id>
+                        <goals>
+                            <goal>prepare-agent</goal>
+                        </goals>
+                    </execution>
+                    <execution>
+                        <id>report</id>
+                        <phase>test</phase>
+                        <goals>
+                            <goal>report</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
```
```suggestion
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/AzureAdxSourceProcessor.java:
```diff
@@ -0,0 +1,192 @@
+public class AzureAdxSourceProcessor extends AbstractProcessor {
+    public static final String ADX_QUERY_ERROR_MESSAGE = "adx.query.error.message";
+    public static final String ADX_EXECUTED_QUERY = "adx.executed.query";
+    public static final Relationship RL_SUCCEEDED = new Relationship.Builder()
```
Using the `RL` prefix is not a standard convention. Although `REL` is sometimes used, recommend eliminating the prefix and changing the name to `SUCCESS`.
```suggestion
    public static final Relationship SUCCESS = new Relationship.Builder()
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/AzureAdxSourceProcessor.java:
```diff
@@ -0,0 +1,192 @@
+            .name(RelationshipStatusEnum.RL_SUCCEEDED.name())
+            .description(RelationshipStatusEnum.RL_SUCCEEDED.getDescription())
+            .build();
+    public static final Relationship RL_FAILED = new Relationship.Builder()
```
```suggestion
    public static final Relationship FAILED = new Relationship.Builder()
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/AzureAdxSourceProcessor.java:
```diff
@@ -0,0 +1,192 @@
+            FlowFile incomingFlowFile = session.get();
+            //incoming connection exists but the incoming flowfile is null
+            if ( incomingFlowFile == null && context.hasNonLoopConnection() ) {
+                getLogger().error("Incoming connection exists but Incoming flow file is null");
```
This status should not be logged as an error, it is a framework concern, so recommend changing the log to an info or removing the log.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/AzureAdxSourceProcessor.java:
```diff
@@ -0,0 +1,192 @@
+        KustoOperationResult kustoOperationResult;
+
+        //checks if this processor has any preceding connection, if yes retrieve
+        if ( context.hasIncomingConnection() ) {
```
```suggestion
        if (context.hasIncomingConnection()) {
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/AzureAdxSourceProcessor.java:
```diff
@@ -0,0 +1,192 @@
+            if ( getLogger().isDebugEnabled() ) {
+                getLogger().debug("Query result {} ", tableData);
+            }
+            try(ByteArrayInputStream bais = new ByteArrayInputStream(json.getBytes())){
+                session.importFrom(bais, outgoingFlowFile);
+            }
```
This type of approach should be changed to use stream the data to an OutputStream using ObjectMapper, instead of holding it all in memory.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/enums/AzureAdxSourceProcessorParamsEnum.java:
```diff
@@ -0,0 +1,42 @@
+ */
+package org.apache.nifi.processors.adx.enums;
+
+public enum AzureAdxSourceProcessorParamsEnum {
```
The `Enum` suffix should be avoided.
```suggestion
public enum AzureAdxSourceProcessorParameter {
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/enums/RelationshipStatusEnum.java:
```diff
@@ -0,0 +1,32 @@
+ */
+package org.apache.nifi.processors.adx.enums;
+
+public enum RelationshipStatusEnum {
```
This enum is probably not necessary as the values can be declared directly in the Relationship definition.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/test/java/org/apache/nifi/processors/adx/AzureAdxSourceProcessorE2ETest.java:
```diff
@@ -0,0 +1,110 @@
+
+    @BeforeEach
+    public void init() throws InitializationException {
+        Assumptions.assumeTrue("true".equalsIgnoreCase(System.getProperty("executeE2ETests")));
```
This approach should be changed to use the `EnabledIfSystemProperty` annotation

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/pom.xml:
```diff
@@ -0,0 +1,154 @@
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-processor-utils</artifactId>
+            <version>1.15.3</version>
+            <exclusions>
+                <exclusion>
+                    <groupId>com.fasterxml.jackson.core</groupId>
+                    <artifactId>jackson-databind</artifactId>
+                </exclusion>
+            </exclusions>
```
This explicit version and exclusion should be removed.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/pom.xml:
```diff
@@ -0,0 +1,154 @@
+        </dependency>
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-processor-utils</artifactId>
```
The `nifi-processor-utils` library is no longer part of the project, so this dependency needs to be replaced with the specific dependency required.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/pom.xml:
```diff
@@ -0,0 +1,154 @@
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+            <version>2.14.0</version>
```
This version is managed at the root level and should be removed.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/pom.xml:
```diff
@@ -0,0 +1,154 @@
+            <artifactId>slf4j-simple</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>jcl-over-slf4j</artifactId>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>log4j-over-slf4j</artifactId>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>jul-to-slf4j</artifactId>
+            <scope>compile</scope>
+        </dependency>
```
```suggestion
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/pom.xml:
```diff
@@ -0,0 +1,154 @@
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <version>5.9.0</version>
```
```suggestion
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/pom.xml:
```diff
@@ -0,0 +1,154 @@
+    </dependencies>
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.jacoco</groupId>
+                <artifactId>jacoco-maven-plugin</artifactId>
+                <version>0.8.6</version>
+                <executions>
+                    <execution>
+                        <id>prepare-agent</id>
+                        <goals>
+                            <goal>prepare-agent</goal>
+                        </goals>
+                    </execution>
+                    <execution>
+                        <id>report</id>
+                        <phase>test</phase>
+                        <goals>
+                            <goal>report</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
```
```suggestion
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/pom.xml:
```diff
@@ -0,0 +1,154 @@
+                    </excludes>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <configuration>
+                    <systemPropertyVariables>
+                        <java.io.tmpdir>${project.build.directory}</java.io.tmpdir>
+                    </systemPropertyVariables>
+                </configuration>
+            </plugin>
```
```suggestion
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/src/main/java/org/apache/nifi/adx/AzureAdxConnectionServiceParamsEnum.java:
```diff
@@ -0,0 +1,43 @@
+ */
+package org.apache.nifi.adx;
+
+public enum AzureAdxConnectionServiceParamsEnum {
```
```suggestion
public enum AzureAdxConnectionServiceParameter {
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/src/main/java/org/apache/nifi/adx/NiFiVersion.java:
```diff
@@ -0,0 +1,51 @@
+ */
+package org.apache.nifi.adx;
+
+import org.apache.commons.lang3.tuple.Pair;import org.slf4j.Logger;
```
```suggestion
import org.apache.commons.lang3.tuple.Pair;
import org.slf4j.Logger;
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/src/main/java/org/apache/nifi/adx/NiFiVersion.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    static {
+        try {
+            Properties props = new Properties();
+            try (InputStream versionFileStream = NiFiVersion.class.getResourceAsStream(VERSION_FILE)) {
+                props.load(versionFileStream);
+                version = props.getProperty("version", version).trim();
+            }
+        } catch (Exception e) {
+            log.warn("Error while loading version:", e);
+        }
```
This approach should be refactored to use `getPackage().getImplementationVersion()` along with adding the information to the module manifest configuration.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/src/main/resources/azure-kusto-nifi-version.properties:
```diff
@@ -0,0 +1,17 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+version=${project.version}
```
As mentioned in the `NiFiVersion` class, this approach should be changed to use the standard Java `getPackage().getImplementationVersion()` approach based on module manifest information.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/pom.xml:
```diff
@@ -0,0 +1,67 @@
+        <module>nifi-adx-api</module>
+    </modules>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.jacoco</groupId>
+                <artifactId>jacoco-maven-plugin</artifactId>
+                <version>0.8.6</version>
+                <executions>
+                    <execution>
+                        <id>prepare-agent</id>
+                        <goals>
+                            <goal>prepare-agent</goal>
+                        </goals>
+                    </execution>
+                    <execution>
+                        <id>report</id>
+                        <phase>test</phase>
+                        <goals>
+                            <goal>report</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
```
```suggestion
```

exceptionfactory reviewed (changes requested):
Thanks for updating the pull request @tanmaya-panda1.

Before evaluating the implementation details, I noted a number of Maven module configuration and stylistic details that should be corrected to follow standard project conventions. Addressing those issues can be done using additional commits to the current branch, and would help clear the way for subsequent reviews.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/AzureAdxSourceProcessor.java:
```diff
@@ -0,0 +1,192 @@
+        @WritesAttribute(attribute = "ADX_QUERY_ERROR_MESSAGE", description = "Azure Data Explorer error message."),
+        @WritesAttribute(attribute = "ADX_EXECUTED_QUERY", description = "Azure Data Explorer executed query.")
+})
+public class AzureAdxSourceProcessor extends AbstractProcessor {
```
Regarding the name, since `Adx` is an acronym that includes `Azure`, recommend renaming the Processor. Also, Processor implementation should not use the word `Processor`.
```suggestion
public class QueryAzureDataExplorer extends AbstractProcessor {
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/QueryAzureDataExplorer.java:
```diff
@@ -0,0 +1,193 @@
+        "This connector can act only as a start of the data pipeline getting data from ADX."+
+        "The queries which can be used further details can be found here https://learn.microsoft.com/en-us/azure/data-explorer/kusto/concepts/querylimits")
+@WritesAttributes({
+        @WritesAttribute(attribute = "ADX_QUERY_ERROR_MESSAGE", description = "Azure Data Explorer error message."),
+        @WritesAttribute(attribute = "ADX_EXECUTED_QUERY", description = "Azure Data Explorer executed query.")
```
FlowFile attribute names generally follow the convention of lowercase with `.` separators. Recommend the following:
```suggestion
        @WritesAttribute(attribute = "adx.query.error.message", description = "Azure Data Explorer error message."),
        @WritesAttribute(attribute = "adx.query.executed", description = "Azure Data Explorer executed query.")
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/QueryAzureDataExplorer.java:
```diff
@@ -0,0 +1,193 @@
+        @WritesAttribute(attribute = "ADX_EXECUTED_QUERY", description = "Azure Data Explorer executed query.")
+})
+public class QueryAzureDataExplorer extends AbstractProcessor {
+    public static final String ADX_QUERY_ERROR_MESSAGE = "adx.query.error.message";
+    public static final String ADX_EXECUTED_QUERY = "adx.executed.query";
```
These two values can be referenced in the `WritesAttribute` annotations.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/QueryAzureDataExplorer.java:
```diff
@@ -0,0 +1,193 @@
+    public static final String RELATIONSHIP_SUCCESS = "SUCCESS";
+
+    public static final String RELATIONSHIP_FAILED = "FAILED";
+    public static final String RELATIONSHIP_FAILED_DESC = "Relationship for failure";
+    public static final String RELATIONSHIP_SUCCESS_DESC = "Relationship for success";
```
These descriptions should not be public static variables, instead they can be specified directly in the Relationship Builder.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/QueryAzureDataExplorer.java:
```diff
@@ -0,0 +1,193 @@
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        this.descriptors = List.of(ADX_SOURCE_SERVICE,DB_NAME,ADX_QUERY);
+        this.relationships = Set.of(SUCCESS,FAILED);
```
These collections can be declared statically and the `init` method can be removed.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/enums/AzureAdxSourceProcessorParameter.java:
```diff
@@ -0,0 +1,42 @@
+package org.apache.nifi.processors.adx.enums;
+
+public enum AzureAdxSourceProcessorParameter {
+    DB_NAME("Database name", "The name of the database where the query will be executed."),
```
Display Names should follow the "Title Case" convention, capitalizing most of the words.
```suggestion
    DB_NAME("Database Name", "The name of the database where the query will be executed."),
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/enums/AzureAdxSourceProcessorParameter.java:
```diff
@@ -0,0 +1,42 @@
+public enum AzureAdxSourceProcessorParameter {
+    DB_NAME("Database name", "The name of the database where the query will be executed."),
+
+    ADX_QUERY("ADX query", "The query which needs to be executed in Azure Data Explorer."),
```
The `ADX` prefix on the property name is not necessary.
```suggestion
    ADX_QUERY("Query", "The query which needs to be executed in Azure Data Explorer."),
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/pom.xml:
```diff
@@ -0,0 +1,118 @@
+            <artifactId>junit-jupiter-api</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-standard-services-api-nar</artifactId>
+            <version>2.0.0-SNAPSHOT</version>
+            <type>nar</type>
+        </dependency>
```
This dependency is not necessary and should be removed.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/pom.xml:
```diff
@@ -0,0 +1,118 @@
+            <version>2.0.0-SNAPSHOT</version>
+            <type>nar</type>
+        </dependency>
+        <dependency>
+            <groupId>com.microsoft.azure.kusto</groupId>
+            <artifactId>kusto-ingest</artifactId>
+            <version>${azure-kusto-java-sdk-version}</version>
+            <exclusions>
+                <exclusion>
+                    <groupId>com.google.code.findbugs</groupId>
+                    <artifactId>jsr305</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-utils</artifactId>
+            <version>2.0.0-SNAPSHOT</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>jcl-over-slf4j</artifactId>
+            <scope>compile</scope>
+        </dependency>
```
Compile dependencies should be listed before test dependencies.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/pom.xml:
```diff
@@ -0,0 +1,118 @@
+        <version>2.0.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>nifi-adx</artifactId>
```
Following general project conventions, the module name should include the word `service` indicating that it contains the Controller Service implementation.
```suggestion
    <artifactId>nifi-adx-service</artifactId>
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/src/main/java/org/apache/nifi/adx/AzureAdxSourceConnectionService.java:
```diff
@@ -0,0 +1,186 @@
+
+@Tags({"Azure", "ADX", "Kusto", "ingest", "azure"})
+@CapabilityDescription("Sends batches of flow file content or stream flow file content to an Azure ADX cluster.")
+@ReadsAttributes({
+        @ReadsAttribute(attribute = "AUTH_STRATEGY", description = "The strategy/method to authenticate against Azure Active Directory, either 'application' or 'managed_identity'."),
+        @ReadsAttribute(attribute = "APP_ID", description = "Specifies Azure application id for accessing the ADX-Cluster."),
+        @ReadsAttribute(attribute = "APP_KEY", description = "Specifies Azure application key for accessing the ADX-Cluster."),
+        @ReadsAttribute(attribute = "APP_TENANT", description = "Azure application tenant for accessing the ADX-Cluster."),
+        @ReadsAttribute(attribute = "CLUSTER_URL", description = "Endpoint of ADX cluster. This is required only when streaming data to ADX cluster is enabled."),
+})
```
Controller Services do not read FlowFile attributes, so these annotations should be removed.
```suggestion
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/src/main/java/org/apache/nifi/adx/AzureAdxSourceConnectionService.java:
```diff
@@ -0,0 +1,186 @@
+})
+public class AzureAdxSourceConnectionService extends AbstractControllerService implements AdxSourceConnectionService {
+
+    private static final String KUSTO_STRATEGY_APPLICATION = "application";
+
+    private static final String KUSTO_STRATEGY_MANAGED_IDENTITY = "managed_identity";
```
These static string values should be replaced with an `enum` that defines the allowed values. The `enum` should implement the `DescribedValue` interface to provide description information.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/src/main/java/org/apache/nifi/adx/AzureAdxSourceConnectionService.java:
```diff
@@ -0,0 +1,186 @@
+
+    @OnStopped
+    public final void onStopped() {
+        if(this.executionClient!=null){
```
```suggestion
        if (this.executionClient != null) {
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/src/main/java/org/apache/nifi/adx/AzureAdxSourceConnectionService.java:
```diff
@@ -0,0 +1,186 @@
+        adxConnectionParams = new ADXConnectionParams();
+        adxConnectionParams.setKustoAuthStrategy(context.getProperty(KUSTO_AUTH_STRATEGY).evaluateAttributeExpressions().getValue());
+        adxConnectionParams.setAppId(context.getProperty(APP_ID).evaluateAttributeExpressions().getValue());
+        adxConnectionParams.setAppKey(context.getProperty(APP_KEY).evaluateAttributeExpressions().getValue());
```
Sensitive properties should not support attribute evaluation.
```suggestion
        adxConnectionParams.setAppKey(context.getProperty(APP_KEY).getValue());
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/src/main/java/org/apache/nifi/adx/AzureAdxSourceConnectionService.java:
```diff
@@ -0,0 +1,186 @@
+            .Builder().name(AzureAdxConnectionServiceParameter.APP_KEY.name())
+            .displayName(AzureAdxConnectionServiceParameter.APP_KEY.getParamDisplayName())
+            .description(AzureAdxConnectionServiceParameter.APP_KEY.getDescription())
+            .required(true)
```
This property should be marked as sensitive.
```suggestion
            .required(true)
            .sensitive(true)
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/src/main/java/org/apache/nifi/adx/NiFiVersion.java:
```diff
@@ -0,0 +1,45 @@
+public class NiFiVersion {
+    public static final String CLIENT_NAME = "Kusto.Nifi";
+
+    public static final Pair<String,String> NIFI_SOURCE = Pair.of("processor", "nifi-source");
```
Is this value necessary?

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/src/main/java/org/apache/nifi/adx/AzureAdxSourceConnectionService.java:
```diff
@@ -0,0 +1,186 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.adx;
```
```suggestion
package org.apache.nifi.adx.service;
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx/src/test/java/org/apache/nifi/adx/MockAzureAdxSourceProcessor.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class MockAzureAdxSourceProcessor extends AbstractProcessor {
```
This class can be removed in favor of a mock created using Mockito.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-api/src/main/java/org/apache/nifi/adx/AdxSourceConnectionService.java:
```diff
@@ -0,0 +1,29 @@
+import org.apache.nifi.annotation.documentation.Tags;
+import org.apache.nifi.controller.ControllerService;
+
+import com.microsoft.azure.kusto.data.Client;
```
This `Client` references places a hard dependency on a specific version of the Kusto library. It looks like the use of the Client in the Processor is limited to return results. Instead of having the interface return the `Client`, it looks like this service interface could be abstract with a `query()` method that returns Data in the form of `List<List<Object>>`. What do you think of that approach? That would allow removing the dependency on the Kusto library from the API module, and make the service more flexible should it be necessary to upgrade the Kusto library for the Controller Service implementation.

exceptionfactory reviewed (changes requested):
Thanks for making the adjustments @tanmaya-panda1, the changes look good.

I went through and provided another round of feedback, most of the recommendations are minor. There is one larger issue related to the interface definition of the `AdxSourceConnectionService`. As noted on the interface itself, it would be much better to define the Controller Service interface in a way that does not depend on the Kusto library. That will allow the Controller Service implementation to change library versions if needed, without necessarily impacting the Processor. Although these two are connected, this provides an important decoupling between Processor and Controller Service.

kustonaut commented:
@exceptionfactory we have addressed the review comments. Request your re-review.

kustonaut commented:
Hi @exceptionfactory, 
Requesting your re-review as we have addressed your comments.

exceptionfactory commented:
> Hi @exceptionfactory, Requesting your re-review as we have addressed your comments.

Thanks @kustonaut and @tanmaya-panda1, I will follow up on this soon.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/QueryAzureDataExplorer.java:
```diff
@@ -0,0 +1,176 @@
+            //execute Query
+            kustoQueryResponse = executeQuery(databaseName,adxQuery);
+            if(!kustoQueryResponse.isError()){
+                try(ByteArrayInputStream bais = new ByteArrayInputStream(objectMapper.writeValueAsBytes(kustoQueryResponse.getTableData()))){
+                    session.importFrom(bais, outgoingFlowFile);
+                }
```
This approach is very memory-intensive and needs to be changed. Serializing the entire result set into a byte array can easily lead to memory exhaustion for large numbers of results. The Kusto ClientFactory also includes a `StreamingClient` that returns an `InputStream`. Based on that option, both the Controller Service interface and this Processor implementation should be changed. The Service should return a streaming query result, which can be transferred directly to the FlowFile OutputStream using one of the methods on ProcessSession.

tanmaya-panda1 replied:
makes sense @exceptionfactory, fixed as advised

exceptionfactory reviewed (changes requested):
Thanks for the recent round of updates @tanmaya-panda1.

The separation of concerns between the Controller Service interface and Processor is on track. In the course of review the Kusto ClientFactory, however, I noticed that there is a `StreamingClient` available that returns results as an `InputStream`. Using that approach would be much more efficient that the current implementation, which reads everything from Data Explorer into objects, then serializes results into memory as a byte array, and then finally writes out the results. Instead, the Controller Service should return a result with an InputStream, which can be copied into a FlowFile OutputStream using the NiFi `StreamUtils.copy()` method together with a `ProcessSession.write()` call.

I noted a number of other naming and stylistic adjustments.

Thanks again for working through the feedback, with the adjustment to use streaming results, this should be nearing completion.

tanmaya-panda1 commented:
@exceptionfactory, Thanks for your review comments, I have fixed and raised the PR. Also I have removed the testcase for kusto query limits exceeded, since the query response limits the results to 500,000 rows when the number of records exceeds more than 500,000 rows. Re-requesting your review.

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/QueryAzureDataExplorer.java:
```diff
@@ -0,0 +1,165 @@
+            session.transfer(outgoingFlowFile, FAILED);
+        }else {
+            // success scenario
+            session.importFrom(kustoQueryResponse.getAdxQueryResponseStream(), outgoingFlowFile);
```
The `ProcessSession.importFrom()` method may not close the provided InputStream, so recommend wrapping this call in try-with-resources:
```suggestion
            try (final InputStream responseStream = kustoQueryResponse.getAdxQueryResponseStream()) {
                session.importFrom(responseStream, outgoingFlowFile);
```

exceptionfactory commented on nifi-nar-bundles/nifi-adx-bundle/nifi-adx-api/src/main/java/org/apache/nifi/adx/model/KustoQueryResponse.java:
```diff
@@ -0,0 +1,50 @@
+
+public class KustoQueryResponse {
+
+    private InputStream adxQueryResponseStream;
+    private boolean isError;
```
Minor naming adjustment recommendation:
```suggestion
    private InputStream responseStream;
    private boolean error;
```

exceptionfactory reviewed (commented):
Thanks for making the adjustments to use the streaming approach @tanmaya-panda1. I noted an issue with closing the response stream, but the general approach looks good.

tanmaya-panda1 commented:
@exceptionfactory : Thanks for reviewing. I have done the changes as requested, can you please have a look at the PR again.

tanmaya-panda1 commented:
@exceptionfactory : A Gentle reminder to the PR, can you please review the changes once.

exceptionfactory reviewed (commented):
Thanks for the latest updates @tanmaya-panda1. This looks closer to completion, although there are still a few stylistic comments that need to be addressed. I should be able to push a commit with some polishing finalization soon.

exceptionfactory reviewed (commented):
@tanmaya-panda1 Thanks for your patience on this review.

After evaluating the structure and dependencies, I moved the QueryAzureDataExplorer Processor and StandardKustoQueryService to the existing `nifi-azure-processors` module under the `nifi-azure-bundle`. The Kusto library has a number of dependencies already present in the `nifi-azure-nar`, so this move significantly reduced the overhead of supporting the new components.

I also replaced the `kusto-ingest` dependency with `kusto-data` to avoid other unnecessary dependencies.

In the course of moving the components, I restructured the property names and unit tests to follow standard project conventions. This eliminated the external `enum` definitions for property names and also streamlined the components themselves.

The only notable behavior change is `QueryAzureDataExplorer` now requires an input FlowFile. This simplifies the input logic significantly and also follows the pattern of other query components. This approach supports populating the Query and Database Name properties from FlowFile attributes if needed.

The QueryAzureDataExplorer Processors successfully writes JSON from the response stream and handles errors as expected.

If you have any questions or concerns on changes, please let me know. I also rebased the pull request on the current main branch, which avoids the merge commit and incorporates more recent updates to Azure dependencies.

If the changes look good, this should be ready to merge soon.

tanmaya-panda1 commented:
Thanks a ton @exceptionfactory for reviewing the PR and implementing the fixes. I think we are good with all changes as suggested. 
But I had a question though on QueryAzureDataExplorer. Since it mandatorily requires a flow file to query into ADX, can this processor become a starting point of a flow where we query into ADX using input query and write the output to a file or some other output processor?

exceptionfactory commented:
> Thanks a ton @exceptionfactory for reviewing the PR and implementing the fixes. I think we are good with all changes as suggested. But I had a question though on QueryAzureDataExplorer. Since it mandatorily requires a flow file to query into ADX, can this processor become a starting point of a flow where we query into ADX using input query and write the output to a file or some other output processor?

You're welcome, and thanks for the follow-up @tanmaya-panda1.

Regarding the required input FlowFile, you are right that the updated approach changes the behavior to require some other starting point Processor, such as GenerateFlowFile, as an input to QueryAzureDataExplorer. This behavior could be restored, but I wanted to clarify the intended use case first. Other "source" Processors, such as the List Processors, use state tracking to provide a query approach that only brings in new information. Without that tracking in this Processor, it would be necessary to right a query with a relative boundary, or using NiFi Expression Language to compute the current time, for example. The same result could be achieved using `GenerateFlowFile` in front of this Processor, but I understand that it introduces an additional step in the data flow. If you can provide some additional details on how you expected to use this Processor as a standalone source, that would be helpful in determining the best path forward.

tanmaya-panda1 commented:
Thanks for your quick reply @exceptionfactory. Our primary focus was to develop the Azure Data Explorer sink for NiFi (PR in pipeline) and this Azure Data Explorer processor as a source for NiFi was done to complete the ecosystem involving Azure Data Explorer and NiFi. So we don't have any particular use case for Azure Data Explorer as a standalone source. 

Maybe we can start with your suggestion to make our processor in line with other source processor and merge this PR. Once we get similar query on supporting standalone source for our processor, we can implement the same.

@exceptionfactory, what do you suggest on this.

exceptionfactory commented:
Thanks for the reply @tanmaya-panda1. With that background, it sounds like merging this PR will provide a helpful baseline for connectivity to Azure Data Explorer. Of course future changes could be implemented as needed. Will plan on merging soon!

kustonaut commented:
[celebrate]     Akshay Dixit reacted to your message:
________________________________
From: exceptionfactory ***@***.***>
Sent: Thursday, June 22, 2023 8:15:59 PM
To: apache/nifi ***@***.***>
Cc: Akshay Dixit ***@***.***>; Mention ***@***.***>
Subject: Re: [apache/nifi] NIFI-8294 Add Query Processor for Azure Data Explorer (PR #7122)


@exceptionfactory approved this pull request.

—
Reply to this email directly, view it on GitHub<https://github.com/apache/nifi/pull/7122#pullrequestreview-1493931033>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ATWGSJDCTOMV6ZNDK3YOW5DXMSRX7ANCNFSM6AAAAAAWS4NZRM>.
You are receiving this because you were mentioned.Message ID: ***@***.***>
